File: src/services/healthKitService.ts
import { Platform } from 'react-native';
import type { WorkoutSession } from '@/types';

// Lazy load the native module to avoid crashes when not available
let HealthKit: typeof import('@kingstinct/react-native-healthkit') | undefined;
let moduleLoadAttempted = false;

function getHealthKit() {
  if (!moduleLoadAttempted && Platform.OS === 'ios') {
    moduleLoadAttempted = true;
    try {
      HealthKit = require('@kingstinct/react-native-healthkit');
    } catch (e) {
      // Module not available - silently continue
    }
  }
  return HealthKit;
}

/**
 * Check if HealthKit is available on this device
 */
export function isHealthKitAvailable(): boolean {
  if (Platform.OS !== 'ios') return false;
  try {
    const hk = getHealthKit();
    return hk?.isHealthDataAvailable?.() ?? false;
  } catch {
    return false;
  }
}

/**
 * Request HealthKit authorization
 * Returns true if authorized, false if denied or unavailable
 */
export async function requestHealthKitAuthorization(): Promise<boolean> {
  if (!isHealthKitAvailable()) {
    return false;
  }

  const hk = getHealthKit();
  if (!hk) return false;

  try {
    // Request write permission for workouts
    const result = await hk.requestAuthorization({
      toShare: [hk.WorkoutTypeIdentifier],
      toRead: [],
    });
    return result;
  } catch (error) {
    return false;
  }
}

/**
 * Check if HealthKit has been authorized
 */
export async function isHealthKitAuthorized(): Promise<boolean> {
  if (!isHealthKitAvailable()) {
    return false;
  }

  const hk = getHealthKit();
  if (!hk) return false;

  try {
    const status = await hk.getRequestStatusForAuthorization({
      toShare: [hk.WorkoutTypeIdentifier],
      toRead: [],
    });
    return status === hk.AuthorizationRequestStatus.unnecessary;
  } catch {
    return false;
  }
}

/**
 * Save a completed workout session to Apple Health
 */
export async function saveWorkoutToHealthKit(
  session: WorkoutSession
): Promise<{ success: boolean; healthKitId?: string; error?: string }> {
  if (!isHealthKitAvailable()) {
    return { success: false, error: 'HealthKit is not available on this device' };
  }

  const hk = getHealthKit();
  if (!hk) {
    return { success: false, error: 'HealthKit module not loaded' };
  }

  try {
    // Calculate workout times and stats
    const startTime = session.startTime || session.date;
    const endTime = session.endTime || new Date().toISOString();
    const totalVolume = calculateWorkoutVolume(session);

    // Build metadata with workout details
    const metadata: Record<string, string | number> = {
      HKExternalUUID: session.id,
    };

    // Add total volume if any sets were completed
    if (totalVolume > 0) {
      metadata['TotalVolumeLbs'] = totalVolume;
    }

    const result = await hk.saveWorkoutSample(
      hk.WorkoutActivityType.traditionalStrengthTraining,
      [],
      new Date(startTime),
      new Date(endTime),
      undefined, // no totals (distance/energy)
      metadata
    );

    return { success: true, healthKitId: result?.uuid };
  } catch (error) {
    return { success: false, error: String(error) };
  }
}

/**
 * Calculate total volume from a workout session
 */
export function calculateWorkoutVolume(session: WorkoutSession): number {
  let totalVolume = 0;

  for (const exercise of session.exercises) {
    for (const set of exercise.sets) {
      if (set.status === 'completed' && set.actualWeight && set.actualReps) {
        totalVolume += set.actualWeight * set.actualReps;
      }
    }
  }

  return totalVolume;
}

File: mobile-apps/ios/LiftMark/Services/HealthKitService.swift
import Foundation
import HealthKit

// MARK: - HealthKit Save Result

struct HealthKitSaveResult {
    let success: Bool
    let healthKitId: String?
    let error: String?
}

// MARK: - HealthKitService

enum HealthKitService {

    private static let healthStore = HKHealthStore()

    // MARK: - Availability

    /// Check if HealthKit is available on this device.
    static func isHealthKitAvailable() -> Bool {
        HKHealthStore.isHealthDataAvailable()
    }

    // MARK: - Authorization

    /// Request write permission for workout data.
    static func requestAuthorization() async -> Bool {
        guard isHealthKitAvailable() else { return false }

        let workoutType = HKObjectType.workoutType()

        do {
            try await healthStore.requestAuthorization(toShare: [workoutType], read: [])
            return true
        } catch {
            Logger.shared.error(.app, "HealthKit authorization failed", error: error)
            return false
        }
    }

    /// Check current authorization status for writing workout data.
    static func isAuthorized() -> Bool {
        guard isHealthKitAvailable() else { return false }

        let workoutType = HKObjectType.workoutType()
        let status = healthStore.authorizationStatus(for: workoutType)
        return status == .sharingAuthorized
    }

    // MARK: - Save Workout

    /// Save a completed workout session to Apple Health.
    static func saveWorkout(_ session: WorkoutSession) async -> HealthKitSaveResult {
        guard isHealthKitAvailable() else {
            return HealthKitSaveResult(success: false, healthKitId: nil, error: "HealthKit is not available on this device")
        }

        do {
            let startDate: Date
            if let startTime = session.startTime {
                startDate = ISO8601DateFormatter().date(from: startTime) ?? Date()
            } else {
                startDate = ISO8601DateFormatter().date(from: session.date) ?? Date()
            }

            let endDate: Date
            if let endTime = session.endTime {
                endDate = ISO8601DateFormatter().date(from: endTime) ?? Date()
            } else {
                endDate = Date()
            }

            let totalVolume = calculateWorkoutVolume(session)

            var metadata: [String: Any] = [
                HKMetadataKeyExternalUUID: session.id
            ]

            if totalVolume > 0 {
                metadata["TotalVolumeLbs"] = totalVolume
            }

            let workout = HKWorkout(
                activityType: .traditionalStrengthTraining,
                start: startDate,
                end: endDate,
                duration: endDate.timeIntervalSince(startDate),
                totalEnergyBurned: nil,
                totalDistance: nil,
                metadata: metadata
            )

            try await healthStore.save(workout)

            return HealthKitSaveResult(success: true, healthKitId: workout.uuid.uuidString, error: nil)
        } catch {
            Logger.shared.error(.app, "Failed to save workout to HealthKit", error: error)
            return HealthKitSaveResult(success: false, healthKitId: nil, error: String(describing: error))
        }
    }

    // MARK: - Volume Calculation

    /// Calculate total workout volume as the sum of (weight x reps) for all completed sets.
    static func calculateWorkoutVolume(_ session: WorkoutSession) -> Double {
        var totalVolume: Double = 0

        for exercise in session.exercises {
            for set in exercise.sets {
                let actual = set.entries.first?.actual
                if set.status == .completed,
                   let weight = actual?.weight?.value,
                   let reps = actual?.reps {
                    totalVolume += weight * Double(reps)
                }
            }
        }

        return totalVolume
    }
}
